---
tags: spillover, virus, database
---

Spillover Meeting, 04/26/2023
==


[![hackmd-github-sync-badge](https://hackmd.io/qYkWEvwITmyWdjI4xFl-Ng/badge)](https://hackmd.io/qYkWEvwITmyWdjI4xFl-Ng)

## Discussion around Virus Taxonomy

- Sounds like the same phenotype/genome taxonomy debates are happening in the virus space as in the microbial space.  
    - ICTV VMR taxonomy relies on a combination of structure, protein, nucleotide and phenotypic information. Groups of viruses are evaluated independently by associated researchers with expertise in each viral group. This is not very automatic/repeatable/etc.
    - virus "name" is below "species" level, but not standardized. These names have been changing over time as well.
- We will rely entirely on genomic sequence, so that we can interconvert between taxonomies, applying ICTV/NCBI/PREDICT information.
    - We can automate + do spot checks, bring in experts if needed
    - LIN taxonomic framework (genomic coordinate system; ~gps coordinates for genomes) is a good option here
        - [LINbase: a web server for genome-based identification of prokaryotes as members of crowdsourced taxa (2020)](https://academic.oup.com/nar/article/48/W1/W529/5813796)
            - this contains a pretty good overview in the intro
        - older, LIN development:
            - [A Proposal for a Genome Similarity-Based Taxonomy for Plant-Pathogenic Bacteria that Is Sufficiently Precise to Reflect Phylogeny, Host Range, and Outbreak Affiliation Applied to Pseudomonas syringae sensu lato as a Proof of Concept (2016)](https://doi.org/10.1094/PHYTO-07-16-0252-R)
            - [A System to Automatically Classify and Name Any Individual Genome-Sequenced Organism Independently of Current Biological Classification and Nomenclature (2014)](https://doi.org/10.1371/journal.pone.0089142)
    - question going forward: how/where do we want to layer in names? E.g. WHO, other organization will want a name to work wiht.

## SpillOver database specifics (33k sequences)
- Some sequences are not in NCBI, but public elsewhere (e.g. Chinese Coronaviruses)
    - Tessa will work with Nistara to get access to these
- ~900 PREDICT Viruses are new, e.g. they may match the ICTV taxonomy/reference genomes at family level or above, but not below.
    - New viruses were ID'd via BLASTn of conserved regions, then calling new viruses when >10% sequence divergence from existing viruses
    - These viruses are accepted/the method has been published:
        - [A Strategy To Estimate Unknown Viral Diversity in Mammals](https://journals.asm.org/doi/10.1128/mBio.00598-13)
        - [Global patterns in coronavirus diversity](https://academic.oup.com/ve/article/3/1/vex012/3866407?login=false)
        - experts Simon Anthony, ?? are connected with this project and could be brought back in to verify genome-based classification/organization system.

## Other notes

[TropMed conference, Oct 17-22 Chicago](https://www.astmh.org/annual-meeting/future-meetings) - suggest Tessa maybe attend to be involved in larger genome/taxonomy discussion

Future work: 
- There is some possibility of expanding to domestic animal viruses and maybe even human viruses (though Jonna has made clear that would require additional funds)

- Can we ID, assess (esp. high-risk) viruses found via genomic surveillance?
    - Jonna: main issue is that datasets are not published in a timely manner. Most are made public with publication, as researchers remain concerned about getting scooped.
    - Need to find a way to shift culture towards immediate data sharing.

[GISAID - Peter Bogner invented persona](https://www.science.org/content/article/invented-persona-behind-key-pandemic-database)